### PR TITLE
Better even_split=False support in gluon.split_data()

### PR DESCRIPTION
**Pull request by JulianSlzr**
## Description ##
When someone uses `even_split=False`, it means they want the data to be processed irrespective of how divisible the amount of data is. Suppose you have 8 GPUs. The old code would:
- Error out if the last batch only had 7 samples
- If you had 64+7 samples, one expects 7 GPUs w/ 9 and 1 GPU w/ 8; however, you got 7 GPUs w/ 8 and 1 GPU w/ 15.

This address both issues.

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Support `size < num_slice` case
- [x] Use `ndarray.split()` when divisible, not just when `even_split=True`
- [x] More evenly distribute splits

## Comments ##
**One may be concerned by returning less than `num_slice` lists when `even_split=False`.** However, this is not a problem idiomatically. E.g., if `num_slice = len(context)`, the expression `zip(context, split)` would truncate to the length of `split <= num_slice`; note that `split_and_load` works as-is after changing `split_data`.

### CODE REVIEW

File: python/mxnet/gluon/utils.py [hunk 1]
```diff
@@ -56,13 +56,10 @@ def split_data(data, num_slice, batch_axis=0, even_split=True):
     Returns
     -------
     list of NDArray
-        Return value is a list even if `num_slice` is 1.
+        Return value is a list even if `num_slice` is 1. When `even_split`
+        is `False` this may be shorter than `num_slice`.
```
Comment (piiswrong) [inline python/mxnet/gluon/utils.py]: Is this a good idea?
when used for multi-gpu, this would cause some gpus to not get data, then gradient will be stale during update

Comment (JulianSlzr) [reply python/mxnet/gluon/utils.py]: See comment above: A user training using `zip(context, slices)` then doing`L.backward()` inside the loop won't have problems.

Also, suppose a user doesn't use this idiom and is unlucky (has `N` samples where `N % batch_size < num_slice`). At worst, MXNet warns them about the stale gradient; plus, it would occur on the very last batch. This is still preferable to doing training/inference just to error on the last batch with a `size < num_slice` message, requiring an ad-hoc workaround.

Comment (piiswrong) [reply python/mxnet/gluon/utils.py]: I think it will fail at step due to stale gradient

Comment (piiswrong) [reply python/mxnet/gluon/utils.py]: It's better to use the discard or rollover option for DataLoader

**Comment by CodingCat** [comment]
Hi, the community has passed to vote about associating the code changes with JIRA (https://lists.apache.org/thread.html/ab22cf0e35f1bce2c3bf3bec2bc5b85a9583a3fe7fd56ba1bbade55f@%3Cdev.mxnet.apache.org%3E)

We have updated the guidelines for contributors in https://cwiki.apache.org/confluence/display/MXNET/Development+Process, please ensure that you have created a JIRA at https://issues.apache.org/jira/projects/MXNET/issues/ to describe your work in this pull request and include the JIRA title in your PR as [MXNET-xxxx] your title where MXNET-xxxx is the JIRA id

Thanks!

**Comment by lupesko** [comment]
@JulianSlzr bouncing.

**Comment by lupesko** [comment]
@piiswrong @sandeep-krishnamurthy @nswamy this PR has not been followed up on for > 1 month. Kindly suggesting to close.
